Soap/Client/Client.php

```diff
@@ -0,0 +1,315 @@
+<?php
+/*
+* This file is part of the CleverAge/ProcessBundle package.
+*
+* Copyright (C) 2017-2018 Clever-Age
+*
+* For the full copyright and license information, please view the LICENSE
+* file that was distributed with this source code.
+*/
+
+namespace CleverAge\ProcessBundle\Soap\Client;
+
+use Psr\Log\LoggerInterface;
+
+/**
+ * Class Client
+ *
+ * @author Madeline Veyrenc <mveyrenc@clever-age.com>
+ */
+class Client implements ClientInterface
+{
+    /**
+     * @var
+     */
+    private $code;
+
+    /**
+     * @var string
+     */
+    private $wsdl;
+
+    /**
+     * @var array
+     */
+    private $options = [];
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * @var \SoapClient
+     */
+    private $soapClient;
+
+    /**
+     * @var string
+     */
+    private $lastRequest;
+
+    /**
+     * @var string
+     */
+    private $lastRequestHeaders;
+
+    /**
+     * @var string
+     */
+    private $lastResponse;
+
+    /**
+     * @var string
+     */
+    private $lastResponseHeaders;
+
+    /**
+     * Client constructor.
+     *
+     * @param LoggerInterface $logger
+     */
+    public function __construct(LoggerInterface $logger)
+    {
+        $this->logger = $logger;
+    }
+
+    /**
+     * {@inheritdoc}
+     * @throws \UnexpectedValueException
+     */
+    public function getCode(): string
+    {
+        if (!$this->code) {
+            throw new \UnexpectedValueException('Client code is not defined');
+        }
+
+        return $this->code;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setCode(?string $code): void
+    {
+        $this->code = $code;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getWsdl(): string
+    {
+        return $this->wsdl;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setWsdl(?string $wsdl): void
+    {
+        $this->wsdl = $wsdl;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getOptions(): array
+    {
+        return $this->options;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setOptions(array $options): void
+    {
+        $this->options = $options;
+    }
+
+    /**
+     * @return LoggerInterface
+     */
+    public function getLogger(): LoggerInterface
+    {
+        return $this->logger;
+    }
+
+    /**
+     * @param LoggerInterface $logger
+     */
+    public function setLogger(LoggerInterface $logger): void
+    {
+        $this->logger = $logger;
+    }
+
+    /**
+     * @return \SoapClient|null
+     */
+    public function getSoapClient(): ?\SoapClient
+    {
+        return $this->soapClient;
```

Comment:
Why not putting the soap client in the constructor ?

Reply:
Some time you need to add some options before creating the client, authentification for exemple.
Most of the time, this class will be override to add authentification, token generation, etc.